providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,734 @@
+                q += key->info->alg[i].pubkey_length_bytes;
+                if (!EVP_PKEY_set_params(key->keys[i], prms))
+                    return 0;
+            }
+            break;
+
+        case PIDX_PKEY_PARAM_PROPERTIES:
+            str = NULL;     /* Force allocation of value */
+            if (!OSSL_PARAM_get_utf8_string(p, &str, SIZE_MAX))
+                return 0;
+            OPENSSL_free(key->propq);
+            key->propq = str;
+            /* Fall through */
+
+        default:
+            if (!key_subalg_set_param(key->info->num_algs, key->keys, prms))
+                return 0;
+            break;
+        }
+    }
+    return 1;
+}
+
+int ossl_hybrid_kmgmt_validate(const void *vctx, int selection, int checktype)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t key_length;
+    unsigned int i;
+    const int pub = (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0;
+    const int priv = (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0;
+
+    if (!ossl_prov_is_running())
+        return 0;
+    if ((key_length = ctx->pubkey_length) == 0)
+        return 0;
+
+    if (pub || priv)
+        for (i = 0; i < ctx->info->num_algs; i++)
+            if ((priv && !EVP_PKEY_private_check(ctx->ctxs[i]))
+                    || (pub && !EVP_PKEY_public_check(ctx->ctxs[i]))
+                    || (pub && priv && !EVP_PKEY_pairwise_check(ctx->ctxs[i])))
+                return 0;
+    return 1;
+}
+
+void ossl_hybrid_pkey_ctx_free(HYBRID_PKEY_CTX *ctx)
+{
+    unsigned int i;
+
+    if (ctx != NULL) {
+        for (i = 0; i < ctx->info->num_algs; i++)
+            EVP_PKEY_CTX_free(ctx->ctxs[i]);
+        OPENSSL_free(ctx->propq);
+        OPENSSL_free(ctx);
+    }
+}
+
+HYBRID_PKEY_CTX *ossl_hybrid_pkey_ctx_alloc(OSSL_LIB_CTX *libctx,
+                                            const HYBRID_ALG_INFO *info)
+{
+    HYBRID_PKEY_CTX *ctx = OPENSSL_zalloc(sizeof(*ctx));
+    unsigned int i;
+
+    if (ctx != NULL) {
+        ctx->libctx = libctx;
+        ctx->info = info;
+        INIT_ACCUMULATE_HYBRID_NUMBERS(ctx);
+        for (i = 0; i < ctx->info->num_algs; i++)
+            ACCUMULATE_HYBRID_NUMBERS(ctx, i);
+    }
+    return ctx;
+}
+
+HYBRID_PKEY_CTX *ossl_hybrid_kmgmt_gen_init(void *provctx,
+                                            int selection,
+                                            const OSSL_PARAM params[],
+                                            const HYBRID_ALG_INFO *info)
+{
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(provctx);
+    HYBRID_PKEY_CTX *ctx;
+    unsigned int i;
+
+    ctx = ossl_hybrid_pkey_ctx_alloc(libctx, info);
+    if (ctx == NULL)
+        return NULL;
+
+    for (i = 0; i < info->num_algs; i++) {
+        EVP_PKEY_CTX_free(ctx->ctxs[i]);
+        ctx->ctxs[i] = EVP_PKEY_CTX_new_from_name(libctx, info->alg[i].name,
+                                                  ctx->propq);
+        if (ctx->ctxs[i] == NULL)
```

Comment:
if i is 1 should this goto err and clean itself up here?

Reply:
I guess so.